### PR TITLE
Feature/evaluate corpus

### LINKED ISSUE #16: As a user, I want a script that generates a CSV containing MT metric scores for each translation in a machine translation corpus, so I can evaluate these machine translations.
Output CSV should have the following fields:
- `tr_id`: The unique identifier of the evaluate translation
- `chrf`: The ChrF score for this translation
- `comet`: The COMET score for this translation

Depends on #14, #15

### PR DESCRIPTION
**Pull request by tanhaow**
**Associated Issue(s):** resolves #16 

### Changes in this PR
_Include all key changes in this pull request_

- Added `evaluate_corpus.py` script that generates CSV files containing MT evaluation metrics (ChrF and COMET scores) for machine translation corpora
- Implemented caching in `metrics.py` to avoid reloading models for each translation, reducing evaluation time from ~9 seconds per translation to ~0.35 seconds per translation
- Suppressed verbose PyTorch Lightning logging messages to provide cleaner output during evaluation

### Notes
- Added `REQUIRED_FIELDS` constant documenting expected input fields in MT corpus records

### Reviewer Checklist
- [x] Run the script on a [sample MT corpus](https://drive.google.com/file/d/1w5C9kQDJPCGURHgXRLEFka5qRbVnVg4q/view?usp=drive_link) to verify CSV output format is correct
- [x] Verify that ChrF and COMET scores are computed correctly and fall within expected ranges
- [x] Verify that progress bar and logging output are clean and informative

### CODE REVIEW

File: src/muse/evaluation/evaluate_corpus.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        "output",
+        type=pathlib.Path,
+        help="Output CSV file with evaluation metrics",
+    )
+    args.add_argument("--verbose", action="store_true", help="Enable verbose output")
+
+    parsed = args.parse_args()
+
+    # Setup logging
+    log_level = logging.DEBUG if parsed.verbose else logging.INFO
```
Comment (laurejt) [inline src/muse/evaluation/evaluate_corpus.py]: Typically, default logging should be set to warning not info, especially when there is a verbose mode.

File: src/muse/evaluation/metrics.py [hunk 2]
```diff
@@ -7,14 +7,38 @@
...
 import os
+import warnings
 
 import evaluate
 import torch
 
-# Suppress verbose HuggingFace logging
+# Suppress verbose HuggingFace and PyTorch Lightning logging
 os.environ["TRANSFORMERS_VERBOSITY"] = "error"
 os.environ["TOKENIZERS_PARALLELISM"] = "false"
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py]: Why do you need to explicitly set this? Were you encountering a warning message?

File: src/muse/evaluation/metrics.py [hunk 3]
```diff
@@ -7,14 +7,38 @@
...
+os.environ["PYTORCH_ENABLE_MPS_FALLBACK"] = "1"
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py]: Why do you need to set this?

File: src/muse/evaluation/metrics.py [hunk 4]
```diff
@@ -7,14 +7,38 @@
...
+os.environ["PYTHONWARNINGS"] = "ignore"
+os.environ["PL_DISABLE_FORK"] = "1"
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py]: Why do you need to set this?

File: src/muse/evaluation/metrics.py [hunk 5]
```diff
@@ -7,14 +7,38 @@
...
 import io
+import logging
 import os
+import warnings
 
 import evaluate
 import torch
 
-# Suppress verbose HuggingFace logging
+# Suppress verbose HuggingFace and PyTorch Lightning logging
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py]: Only a few of these environment variables are logging related. All others must be documented since they impact the execution of pytorch & huggingface

File: src/muse/evaluation/evaluate_corpus.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+import pathlib
+import sys
+
+import orjsonl
+from tqdm import tqdm
+
+from muse.evaluation.metrics import compute_chrf, compute_comet
+
+# Required fields in input machine translation corpus records
+REQUIRED_FIELDS = ["tr_id", "src_text", "ref_text", "tr_text"]
```
Comment (laurejt) [inline src/muse/evaluation/evaluate_corpus.py]: This is fine, but unnecessary. There's no need to validate input we control for experimental code.

File: src/muse/evaluation/metrics.py [hunk 7]
```diff
@@ -7,14 +7,38 @@
...
+    logging.getLogger(logger_name).setLevel(logging.ERROR)
+    logging.getLogger(logger_name).propagate = False
+
+warnings.filterwarnings("ignore")
+
+# Cache for loaded metrics to avoid reloading models
+LOADED_METRICS = {
+    "chrf": None,
+    "comet": None,
+}
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py:33]: How much RAM does this require? This will be important especially worth noting when CometKiwi is added.

File: src/muse/evaluation/metrics.py [hunk 8]
```diff
@@ -28,7 +52,11 @@ def compute_chrf(
     Returns a float in the range [0, 100], where 0 indicates no match and 100
     indicates a perfect match.
     """
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py:46]: In retrospect, the range of ChrF is odd since it typically ranges between 0-1 since it's an f-score. This must have something to do with the implementation that huggingface-evaluate uses. 

Let's correct this by normalizing the scores to 0-1 (i.e., divide by 100) and update the relevant documentation.

File: src/muse/evaluation/metrics.py [hunk 9]
```diff
@@ -53,15 +81,23 @@ def compute_comet(
...
+    if LOADED_METRICS["comet"] is None:
+        # Suppress stdout/stderr during model loading
+        with (
+            contextlib.redirect_stdout(io.StringIO()),
+            contextlib.redirect_stderr(io.StringIO()),
+        ):
+            LOADED_METRICS["comet"] = evaluate.load("comet")
+
+    comet_metric = LOADED_METRICS["comet"]
+    gpus = 1 if (torch.cuda.is_available() or torch.backends.mps.is_available()) else 0
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py:87]: This is fine for now, we may want to update this logic if we plan on running this on a cluster (i.e., make it possible to specify the use of gpu or cpu-only regardless of what the device has available).

**Review by laurejt — CHANGES_REQUESTED** [review]
Overall, this looks good and only needs a few small changes. Requested changes:
- Document why pytorch and huggingface-related enivornment variables are set with `metrics.py`. Not all of these are related to logging.
- As a small add-on (since you're already updating `metrics.py`), please normalize the chrF score so its range is 0-1.

File: src/muse/evaluation/metrics.py (hunk 2 above)
Comment (tanhaow) [reply src/muse/evaluation/metrics.py]: There was a lot of INFO messages suggesting LitLogger, which made the output long and messy:

```
INFO: 💡 Tip: For seamless cloud logging and experiment tracking, try installing [litlogger](https://pypi.org/project/litlogger/) to enable LitLogger, which logs metrics and artifacts automatically to the Lightning Experiments platform.
INFO: 💡 Tip: For seamless cloud uploads and versioning, try installing [litmodels](https://pypi.org/project/litmodels/) to enable LitModelCheckpoint, which syncs automatically with the Lightning model registry.
```

File: src/muse/evaluation/metrics.py (hunk 3 above)
Comment (tanhaow) [reply src/muse/evaluation/metrics.py]: It's for enabling fallback for unsupported MPS operations. But I can remove this.

File: src/muse/evaluation/metrics.py (hunk 2 above)
Comment (laurejt) [reply src/muse/evaluation/metrics.py]: I see this pointing to the bigger issue that the code assumes that INFO is the default non-verbose level, which is not standard.

File: src/muse/evaluation/metrics.py (hunk 4 above)
Comment (tanhaow) [reply src/muse/evaluation/metrics.py]: It's recommended by HuggingFace to prevent tokenizer deadlocks when used with PyTorch's parallel processing during COMET evaluation. I will add a note in the code

File: src/muse/evaluation/metrics.py (hunk 2 above)
Comment (tanhaow) [reply src/muse/evaluation/metrics.py]: You are right. I have set the default logging level to WARNING. Let me also remove some of these logging related code

Comment (tanhaow) [reply src/muse/evaluation/metrics.py]: So, the thing is, the PyTorch Lightning messages are still showing after setting the default logging to WARN. I did some investigation and found it's because PyTorch Lightning uses its own internal logging system that bypasses Python's logging module... So we actually need to keep the stderr `contextlib` redirects. But I will simplify it into a minimal version

**Comment by tanhaow** [comment]
Thanks for your review, @laurejt ! I've made the following changes:

- Change default logging level to WARNING
- Simplify and document environment variable configuration
- Normalize ChrF scores to 0-1 range (-- good catch!)
- Document RAM requirements for COMET caching
- Remove REQUIRED_FIELDS constant

File: src/muse/evaluation/metrics.py (hunk 2 above)
Comment (laurejt) [reply src/muse/evaluation/metrics.py]: Thanks for looking into this!

File: src/muse/evaluation/metrics.py (hunk 3 above)
Comment (laurejt) [reply src/muse/evaluation/metrics.py]: There's no need to remove this, just document it. It's worth documenting that this library supports MPS / Apple Silicon whereas we have not done that with other parts of the pipeline / package.

**Review by laurejt — COMMENTED** [review]
When I run `evaluate_corpus.py` none of the pytorch lighnining INFO logging messages are suppressed.

**Comment by tanhaow** [comment]
PyTorch Lightning logging is now suppressed. Now only 1 INFO message appears on first evaluation (checkpoint upgrade notice) instead of 6 messages per translation.

**Review by laurejt — COMMENTED** [review]
Pytorch lightning does use Python logging, but I think this is a case of not being specific enough with the logger name. I was able to suppress the logging messages with the following two lines:

```
logging.getLogger("pytorch_lightning.utilities.rank_zero").setLevel(logging.WARNING)
logging.getLogger("pytorch_lightning.utilities.migration").setLevel(logging.WARNING)
```

Try simplifying the code to this and verify that this works for your local dev environment as well.

In case you're curious, I figured out the particular module names by searching for the portions of the INFO logs within the [pytorch-lightning repo](https://github.com/Lightning-AI/pytorch-lightning/).

File: src/muse/evaluation/metrics.py [hunk 10]
```diff
@@ -86,17 +92,26 @@ def compute_comet(
...
-        contextlib.redirect_stderr(io.StringIO()),
-    ):
+    # Suppress PyTorch Lightning INFO messages by redirecting stderr at file descriptor level
+    # This is necessary because PyTorch Lightning bypasses Python's logging system
+    _stderr_fd = sys.stderr.fileno()
+    _original_stderr_fd = os.dup(_stderr_fd)
+    _devnull_fd = os.open(os.devnull, os.O_WRONLY)
+    os.dup2(_devnull_fd, _stderr_fd)
+
+    try:
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py]: This is a lot of complexity for what we're trying to do, see if my suggestion works instead.

File: src/muse/evaluation/metrics.py [hunk 11]
```diff
@@ -86,17 +92,26 @@ def compute_comet(
...
             predictions=[tr_text],
             references=[ref_text],
             sources=[src_text],
             gpus=gpus,
         )
+    finally:
+        # Restore original stderr
+        os.dup2(_original_stderr_fd, _stderr_fd)
+        os.close(_original_stderr_fd)
+        os.close(_devnull_fd)
```
Comment (laurejt) [inline src/muse/evaluation/metrics.py]: This is a lot of complexity for what we're trying to do, see if my suggestion works instead.

**Comment by tanhaow** [comment]
> Pytorch lightning does use Python logging, but I think this is a case of not being specific enough with the logger name. I was able to suppress the logging messages with the following two lines:
> 
> ```
> logging.getLogger("pytorch_lightning.utilities.rank_zero").setLevel(logging.WARNING)
> logging.getLogger("pytorch_lightning.utilities.migration").setLevel(logging.WARNING)
> ```
> 
> Try simplifying the code to this and verify that this works for your local dev environment as well.
> 
> In case you're curious, I figured out the particular module names by searching for the portions of the INFO logs within the [pytorch-lightning repo](https://github.com/Lightning-AI/pytorch-lightning/).

This does look like a much more elegant solution!! Ugh you're so smart

**Comment by tanhaow** [comment]
@laurejt It works perfectly. Thank you for digging out this solution!!

**Review by laurejt — APPROVED** [review]
🚀
